### PR TITLE
huc6270.cpp: Implement 4 color background, 16 bit host interface, Add hardware notes

### PR DESCRIPTION
**Pull request by cam900**
according from manual.
Fix spacing

### CODE REVIEW

File: src/devices/video/huc6270.h [hunk 1]
```diff
@@ -21,15 +286,24 @@ class huc6270_device : public device_t
...
+	// VD0-7, SPBG output 
 	u16 next_pixel();
 	inline u16 time_until_next_event()
 	{
 		return m_horz_to_go * 8 + m_horz_steps;
 	}
 
+	// /VSYNC pin related
 	DECLARE_WRITE_LINE_MEMBER( vsync_changed );
+	// /HSYNC pin related
```
Comment (angelosa) [inline src/devices/video/huc6270.h:306]: Function signatures are clear enough to define those as pins, commentary is superfluous.

File: src/devices/video/huc6270.h [hunk 2]
```diff
@@ -21,15 +286,24 @@ class huc6270_device : public device_t
...
 	auto irq() { return m_irq_changed_cb.bind(); }
 
+	// 8 bit I/O
 	u8 read(offs_t offset);
 	void write(offs_t offset, u8 data);
+	// TODO: 16 bit I/O used?
+	/*
+	u16 read16(offs_t offset);
+	void write16(offs_t offset, u16 data);
+	*/
```
Comment (angelosa) [inline src/devices/video/huc6270.h:296]: Internally HuC6270 **is** 16-bit, it's just the interface with PCE HuC6280 that is 8-bit, at least with normal opcodes (no idea with ST0/ST1/ST2 opcodes). And even then, PCFX V810 can do definitely go up to 32, so the whole point is that register access and internal state should be isolated into a single `io_space` that sets up `mem_mask` accordingly and nuke the dual switch-case.

TL;DR both notes are superfluous, plus your commented signatures are wrong (missing mem_mask)

File: src/devices/video/huc6270.h [hunk 3]
```diff
@@ -4,6 +4,271 @@
...
+	- xxxx ---- Sprite Color (when VD0-VD3 != 0)
+	- ---- x--- SG3
+	- ---- -x-- SG2
+	- ---- --x- SG1
+	- ---- ---x SG0
+	- ---- 0000 Transparency, bit 7-4 is forced to 0
+
+	Border:
+	1 0000 0000 Border color, during blanking period
+
```
Comment (angelosa) [inline src/devices/video/huc6270.h:271]: Totally not a fan of having register implementation details in header for visibility reasons, every description should really go near the relevant register fn or at top of .cpp header instead.

**Review by angelosa — COMMENTED** [review]
Again, where did you test "4 color background"?

File: src/devices/video/huc6270.cpp [hunk 4]
```diff
@@ -662,169 +933,85 @@ void huc6270_device::write(offs_t offset, u8 data)
...
-				case CR:        /* control register MSB */
-					m_cr = ( m_cr & 0x00FF ) | ( data << 8 );
+				case CR:        /* control register */
+					COMBINE_DATA(&m_cr);
 					break;
 
-				case RCR:       /* raster compare register MSB */
-					m_rcr = ( m_rcr & 0x00FF ) | ( ( data & 0x03 ) << 8 );
+				case RCR:       /* raster compare register */
+					m_rcr = COMBINE_DATA(&m_rcr) & 0x3ff;
```
Comment (angelosa) [inline src/devices/video/huc6270.cpp:952]: > error: operation on ‘((huc6270_device*)this)->huc6270_device::m_rcr’ may be undefined [-Werror=sequence-point]

**Review by angelosa — CHANGES_REQUESTED** [review]

**Comment by cam900** [comment]
4 color background is confirmed from developer manual but it's seems like rarely used.

**Comment by angelosa** [comment]
> 4 color background is confirmed from developer manual but it's seems like rarely used.

That means you didn't do your homework. You can either:
1. Find a SW reference that matches your handling;
2. Make a test and run it on HW where it matches your handling. Your [pceas2 source code](https://www.romhacking.net/utilities/839/) can be published at `src/mame/etc/whateverfoo.asm` for now, lacking a better infrastructure for the time being;

Manual documentation note isn't enough proof, also to me it looks like that CG0 or CG1 upper bits can really be 1 rather than 0.
Unlikely but not impossible.

**Comment by cam900** [comment]
result of CG0 only, CG1 only, Both; source from https://www.chrismcovell.com/PCEdev/

**Comment by angelosa** [comment]
Steps to reproduce? Assuming this is [tutorial06_05.pce](https://www.chrismcovell.com/PCEdev/tutorial06.zip) we're talking about I don't see attempts to enable mode 4, or being done so in [06a](https://youtu.be/t4b_kuOL63o) or [06b](https://youtu.be/JYZQcj5zrZE), admittedly I've only seen excerpts to those.

**Comment by cam900** [comment]
> Steps to reproduce? Assuming this is [tutorial06_05.pce](https://www.chrismcovell.com/PCEdev/tutorial06.zip) we're talking about I don't see attempts to enable mode 4, or being done so in [06a](https://youtu.be/t4b_kuOL63o) or [06b](https://youtu.be/JYZQcj5zrZE), admittedly I've only seen excerpts to those.

with source code modification, line 77 of INCLUDE/startup.asm controls MWR.

**Comment by cam900** [comment]
I close this PR and will reopen when working test code in real hardware is written.
